### PR TITLE
[omni, model] feat: add fused MoE forward and EP support for Qwen3-Omni-MoE thinker

### PR DESCRIPTION
**Pull request by TimYangst**
### What does this PR do?

> Enables fused MoE forward and Expert Parallelism (EP) for the Qwen3-Omni-MoE **thinker** text model, following the same pattern as Qwen3-VL-MoE. The talker is left as a TODO.
>
> The key change is replacing the HF `nn.ModuleList`-based per-expert MLP with a new `Qwen3OmniMoeThinkerTextExperts` module that stores weights as stacked 3-D tensors `(num_experts, out, in)`, compatible with both the fused grouped-GEMM kernel and FSDP2 EP sharding.

Note: the weights of the Qwen3OmniMoe need to merged by the script: https://github.com/ByteDance-Seed/VeOmni/pull/495

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: https://github.com/ByteDance-Seed/VeOmni/pulls?q=fused+moe
- [ ] Format the PR title as `[{modules}] {type}: {description}`

### Test

Validated by running `tests/models/omni/test_qwen3_omni_moe.py` — HF eager vs VeOmni eager vs VeOmni fused loss/gnorm comparison passes.

### API and Usage Example

Set `_moe_implementation = "fused"` in the thinker config to enable the fused path:

```python
config._moe_implementation = "fused"
model = build_foundation_model(config_path=..., ...)
```

No API changes to the training scripts.

### Design & Code Changes

**`modeling_qwen3_omni_moe.py`**

- Add `Qwen3OmniMoeThinkerTextExperts` — replaces `nn.ModuleList` with stacked `nn.Parameter` tensors (`gate_proj`, `up_proj`, `down_proj` each of shape `(num_experts, out, in)`). Supports:
  - `fused_forward`: calls `fused_moe_forward` kernel (training, `_moe_implementation="fused"`)
  - `eager_forward`: loop-per-expert path (inference / eager mode)
  - `_load_from_state_dict`: transparently converts HF per-expert format (`experts.{i}.{proj}.weight`) to stacked format on load, so both original HF and pre-merged checkpoints are accepted
- Patch `Qwen3OmniMoeThinkerTextSparseMoeBlock.__init__` to instantiate `Qwen3OmniMoeThinkerTextExperts` instead of `nn.ModuleList`
- Patch `Qwen3OmniMoeThinkerTextSparseMoeBlock.forward` to pass `routing_weights` and `selected_experts` directly to the experts module
- Patch `Qwen3OmniMoeThinkerForConditionalGeneration.__init__` to propagate `config._moe_implementation` → `config.text_config._moe_implementation` before the text model is built; guarded against multiple calls to `apply_veomni_qwen3_omni_moe_patch()` to prevent infinite recursion

**`parallel_plan.py`**

- Update EP plan key paths from `experts.*.{proj}.weight` (per-expert ModuleList paths) to `experts.{proj}` (stacked Parameter paths), with `Shard(0)` on the expert dimension
- TODO comment added for talker EP plan

### Checklist Before Submitting

- [ ] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md?plain=1#L22): `make commit && make style && make quality`
- [ ] Add / Update [the documentation](https://github.com/ByteDance-Seed/VeOmni/blob/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/ByteDance-Seed/VeOmni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: covered by existing `test_qwen3_omni_moe.py`

### CODE REVIEW

File: veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py [hunk 1]
```diff
@@ -4515,6 +4516,180 @@ def forward(
...
+        """Eager (loop-per-expert) forward – used for inference / non-fused path."""
+        num_tokens, hidden_dim = hidden_states.shape
+        final_hidden_states = torch.zeros_like(hidden_states)
+
+        expert_mask = torch.nn.functional.one_hot(selected_experts, num_classes=self.num_experts).permute(2, 1, 0)
+        expert_hit = torch.greater(expert_mask.sum(dim=(-1, -2)), 0).nonzero()
+        for expert_idx in expert_hit:
+            # expert_idx is shape [1] from nonzero(); use .item() for scalar indexing into 3-D weights
+            e = expert_idx.item()
+            idx, top_x = torch.where(expert_mask[e].squeeze(0))
```
[the remark this answers is not in the document]
Comment (TimYangst) [reply veomni/models/transformers/qwen3_omni_moe/modeling_qwen3_omni_moe.py]: Done

**Comment by TimYangst** [comment]
Combined with pr: https://github.com/ByteDance-Seed/VeOmni/pull/495
